contracts/ethregistrar/ETHRegistrarController.sol

```diff
@@ -37,197 +27,296 @@ contract ETHRegistrarController is
     function register(
-        string calldata name,
-        address owner,
-        uint256 duration,
-        bytes32 secret,
-        address resolver,
-        bytes[] calldata data,
-        bool reverseRecord,
-        uint16 ownerControlledFuses
+        Registration calldata registration
     ) public payable override {
-        IPriceOracle.Price memory price = rentPrice(name, duration);
-        if (msg.value < price.base + price.premium) {
-            revert InsufficientValue();
-        }
-
-        _consumeCommitment(
-            name,
-            duration,
-            makeCommitment(
-                name,
-                owner,
-                duration,
-                secret,
-                resolver,
-                data,
-                reverseRecord,
-                ownerControlledFuses
-            )
-        );
-
-        uint256 expires = nameWrapper.registerAndWrapETH2LD(
-            name,
-            owner,
-            duration,
-            resolver,
-            ownerControlledFuses
+        bytes32 labelhash = keccak256(bytes(registration.label));
+        IPriceOracle.Price memory price = _rentPrice(
+            registration.label,
+            labelhash,
+            registration.duration
         );
+        uint256 totalPrice = price.base + price.premium;
+        if (msg.value < totalPrice) revert InsufficientValue();
 
-        if (data.length > 0) {
-            _setRecords(resolver, keccak256(bytes(name)), data);
-        }
+        if (!_available(registration.label, labelhash))
+            revert NameNotAvailable(registration.label);
 
-        if (reverseRecord) {
-            _setReverseRecord(name, resolver, msg.sender);
-        }
+        _register(registration, labelhash, price);
 
-        emit NameRegistered(
-            name,
-            keccak256(bytes(name)),
-            owner,
-            price.base,
-            price.premium,
-            expires
-        );
+        if (msg.value > totalPrice)
+            payable(msg.sender).transfer(msg.value - totalPrice);
+    }
 
-        if (msg.value > (price.base + price.premium)) {
-            payable(msg.sender).transfer(
-                msg.value - (price.base + price.premium)
-            );
-        }
+    /// @notice Registers a name without payment, only for owner.
+    /// @dev Bypasses value and validity check (i.e. no payment, <3 chars).
+    ///
+    /// @param registration The registration to register.
+    /// @param registration.label The label of the name.
+    /// @param registration.owner The owner of the name.
+    /// @param registration.duration The duration of the registration.
+    /// @param registration.resolver The resolver for the name.
+    /// @param registration.data The data for the name.
+    /// @param registration.reverseRecord The reverse record for the name.
+    /// @param registration.referrer The referrer of the registration.
+    function registerWithoutPayment(
```

Comment:
This function also shouldn't require a valid commitment, since it's going to be onchain for a week before it's executed!